src/repr/src/row.proto

```diff
@@ -0,0 +1,135 @@
+        // id (fill in the new field on write, but check if either field is set
+        // on read). However, once a 1-byte id is used, it's gone forever, so
+        // we're convervative in handing them out.
+        //
+        // Of these, I'd guess Timestamp and UUID are probably the first ones
+        // we'd bless followed by Date and Time.
+        ProtoDate date = 16;
+        ProtoTime time = 17;
+        ProtoTimestamp timestamp = 18;
+        ProtoInterval interval = 19;
+        ProtoArray array = 20;
+        ProtoRow list = 21;
+        ProtoDict dict = 22;
+        ProtoNumeric numeric = 23;
+        bytes uuid = 24;
+    }
+}
+
+enum ProtoDatumOther {
+    // It's generally good practice to make id 0 (the default if the field is
+    // unset) in proto enums be an Unknown sentinel. This allows for
+    // distinguishing between unset and any of the enum variants.
+    //
+    // This enum is initially used only in a oneof, which means we can
+    // distinguish unset without this sentinel. But stick one in here anyway,
+    // in case this enum gets used somewhere else in the future.
+    Unknown = 0;
+    Null = 1;
+    False = 2;
+    True = 3;
+    JsonNull = 4;
+    Dummy = 5;
+}
+
+message ProtoDate {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+}
+
+message ProtoTime {
+    // The number of seconds since midnight
+    uint32 secs = 1;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 2;
+}
+
+message ProtoTimestamp {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+    // The number of seconds since midnight
+    uint32 secs = 3;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 4;
+    // If true, this timestamp is in UTC. If false, this timestamp is zoneless.
+    bool is_tz = 5;
+}
+
+message ProtoInterval {
+    // A possibly negative number of months.
+    int32 months = 1;
+    // A 128-bit nanosecond count, represented as a 64-bit lo and hi.
+    //
+    // TODO: Why do our in-mem intervals use this representation? I've seen
+    // month/day/nanos in the past becuase a day is not always the same number
+    // of nanos.
+    int64 duration_lo = 2;
+    int64 duration_hi = 3;
+}
+
+message ProtoArray {
+    // All array elements flattened into 1 dimension, encoded in row-major
+    // order.
+    ProtoRow elements = 1;
+    // TODO: This is a direct translation of our in-mem representation of array
+    // dimensions. Does anyone understand it well enough to document it?
+    repeated ProtoArrayDimension dims = 2;
+}
+
+message ProtoArrayDimension {
+    uint64 lower_bound = 1;
+    uint64 length = 2;
+}
+
+message ProtoDict {
+    map<string, ProtoDatum> elements = 1;
```

Comment:
This doesn't look safe because it'll codegen into a HashMap, and we require that maps are stored in ascending order. I think probably we want a repeated set of (string, ProtoDatum) pairs instead?

Reply:
Good call will do